File: desktop/src/workspaces.ts
/**
 * The projects Sailor has been opened in. **THE FIELD NAMES ARE THE CONTRACT**
 * written in `src-tauri/src/workspaces.rs`: whoever changes one changes both,
 * and `workspaces.test.ts` compares the two so a rename cannot pass in silence.
 */
import { invoker } from "./engine";

/**
 * Outside the native shell every call fails rather than pretending. A list of
 * projects is read from the home on disk: a browser cannot be shown a
 * plausible version of it, and one would teach something untrue.
 */
function ask<T>(command: string, args?: Record<string, unknown>): Promise<T> {
  const invoke = invoker();
  if (!invoke) return Promise.reject(new Error("outside the desktop shell: no home to read"));
  return invoke<T>(command, args);
}

/** Whether the marker is still where the project was left. */
export type Standing = "declared" | "gone";

export interface Project {
  root: string;
  name: string;
  first_seen: number;
  last_seen: number;
  standing: Standing;
  /** The project the window is standing in, if it is one of these. */
  current: boolean;
}

/** What a project declares about itself in its own `sailor.json`. */
export interface Declaration {
  name: string;
  rules: string[];
  checks: Record<string, string>;
  equipment: string | null;
}

/**
 * The projects, most recently opened first.
 *
 * An empty list is an answer, not a failure: nobody has declared a project yet,
 * and the screen that shows it says how to declare one.
 */
export function projects(): Promise<Project[]> {
  return ask<Project[]>("workspaces");
}

/** What one project declares. Read when someone looks at it, not per row. */
export function declarationOf(root: string): Promise<Declaration> {
  return ask<Declaration>("workspace_declaration", { root });
}

/**
 * How long ago, in the words a person uses. Not a date: a list of projects is
 * scanned for «which one was I in», and «3 days ago» answers that where
 * «2026-08-30» has to be worked out.
 */
export function since(seconds: number, now: number): string {
  const gap = Math.max(0, now - seconds);
  if (gap < 90) return "just now";
  const minutes = Math.round(gap / 60);
  if (minutes < 60) return `${minutes} min ago`;
  const hours = Math.round(minutes / 60);
  if (hours < 24) return `${hours} h ago`;
  const days = Math.round(hours / 24);
  if (days < 30) return `${days} d ago`;
  return `${Math.round(days / 30)} mo ago`;
}

File: desktop/src-tauri/src/workspaces.rs
//! The projects Sailor has been opened in, as the window asks for them.
//!
//! **THE REGISTER LIVES IN `flow::workspace`, NOT HERE.** This reads it and
//! reshapes it for the canvas; what a project is, and when it counts as gone,
//! belongs to the engine — the command line answers with the same code, and
//! two answers that drift apart beat no answer nobody likes.

use serde::Serialize;

/// A project as the canvas receives it. The field names are the contract
/// written in `desktop/src/workspaces.ts`: whoever changes one changes both.
#[derive(Serialize)]
pub(crate) struct Project {
    root: String,
    name: String,
    first_seen: i64,
    last_seen: i64,
    /// `declared` when the marker is still where it was left, `gone` when it is
    /// not. Read at the moment of asking: a stored answer would be a second
    /// truth to keep aligned with the disk.
    standing: &'static str,
    /// Whether this is the project the window is standing in. The list is
    /// worth little without it — «switch project» needs to know what not to
    /// switch to.
    current: bool,
}

/// The projects, most recently opened first.
///
/// **AN EMPTY HOUSE IS AN EMPTY LIST, NOT A FAILURE.** Nobody has opened a
/// project yet is a complete answer, and the canvas has a card for it; an error
/// there would make a first run look like a broken install.
#[tauri::command]
pub(crate) fn workspaces() -> Result<Vec<Project>, String> {
    let home =
        ledger::sailor_home().ok_or_else(|| "no house to read: HOME is not set".to_owned())?;
    let here = std::env::current_dir()
        .ok()
        .and_then(|from| flow::workspace::find_root(&from));

    Ok(flow::workspace::known_in(&home)?
        .into_iter()
        .map(|known| Project {
            standing: match flow::workspace::standing_of(&known) {
                flow::workspace::Standing::Declared => "declared",
                flow::workspace::Standing::Gone => "gone",
            },
            current: here.as_deref() == Some(known.root.as_path()),
            root: known.root.to_string_lossy().into_owned(),
            name: known.name,
            first_seen: known.first_seen,
            last_seen: known.last_seen,
        })
        .collect())
}

/// What a project declares about itself, read from its own marker.
///
/// Separate from the list on purpose: the list is read on every paint, and a
/// declaration is read when someone looks at one project. Reading every marker
/// to draw a list would touch the disk once per row for something nobody asked.
#[tauri::command]
pub(crate) fn workspace_declaration(root: String) -> Result<serde_json::Value, String> {
    let declared = flow::workspace::declaration_at(std::path::Path::new(&root))?;
    serde_json::to_value(&declared).map_err(|error| error.to_string())
}
